### ISSUE TITLE
postgresql-11 build failure: clang-6.0 / llvm-6.0-dev but it is not installable

### ISSUE THREAD

**Issue by tianon** [docker-library/postgres#484]
See https://travis-ci.org/docker-library/postgres/jobs/418437899#L1197-L1210:

```console
+ apt-get build-dep -y postgresql-common pgdg-keyring postgresql-11=11~beta3-1.pgdg90+2
Reading package lists...
Building dependency tree...
Reading state information...
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:
The following packages have unmet dependencies:
 builddeps:postgresql-11 : Depends: clang-6.0 but it is not installable
                           Depends: llvm-6.0-dev but it is not installable
E: Unable to correct problems, you have held broken packages.
```

(Noted at https://github.com/docker-library/postgres/pull/481#issuecomment-414476328.)

The following upstream commits are relevant:

- https://salsa.debian.org/postgresql/postgresql/commit/883957c4d61179c4bf42fe8680e1389e72affc59
- https://salsa.debian.org/postgresql/postgresql/commit/23b9d5aaa45218dcb9a6fb91d1ad547f1fb3362a

```diff
diff --git a/debian/control b/debian/control
index da86ea8..7087f1c 100644
--- a/debian/control
+++ b/debian/control
@@ -9,7 +9,7 @@ Uploaders:
 Standards-Version: 4.1.4
 Build-Depends:
  bison,
- clang [amd64 i386],
+ clang-6.0 [amd64 i386],
  debhelper (>= 9~),
  dh-exec (>= 0.13~),
  docbook-xml,
@@ -30,7 +30,7 @@ Build-Depends:
  libxml2-dev,
  libxml2-utils,
  libxslt1-dev,
- llvm-dev [amd64 i386],
+ llvm-6.0-dev [amd64 i386],
  perl (>= 5.8),
  pkg-config,
  python-dev,
@@ -195,9 +195,9 @@ Package: postgresql-server-dev-11
 Architecture: any
 Section: libdevel
 Depends:
- clang [amd64 i386],
+ clang-6.0 [amd64 i386],
  libpq-dev (>= 11~~),
- llvm-dev [amd64 i386],
+ llvm-6.0-dev [amd64 i386],
  postgresql-common (>= 142~),
  ${misc:Depends},
  ${shlibs:Depends}
```

LLVM/Clang support was added for PostgreSQL 11+, and is used for JIT compilation of certain queries (see https://github.com/docker-library/postgres/issues/475 for a little more discussion around that).

**Comment by tianon** [comment]
@ChristophBerg would you be willing to comment on my proposed solution here? :pray:

I'm thinking we'll probably just add `stretch-backports` to 11+ for now on arches where we have to build from source (to pull in `clang-6.0`), but if there's a better / more-correct solution I'm all for it. :sweat_smile: :heart:

**Comment by ChristophBerg** [comment]
For apt.postgresql.org, I copied the bpo packages to our repository (amd64/i386/ppc64 (even if the latter is unused yet)).
If you want other (or no) clang/llvm versions, try https://git.postgresql.org/gitweb/?p=pgapt.git;a=blob;f=jenkins/generate-pgdg-source#l148

The problem with versions earlier than 6.0 is that they don't cope with C++ exceptions, so certain PostGIS queries make the server crash.

I plan to go to LLVM 7 asap because it will support non-x86 architectures (problem fixed after 6.0.1).

**Comment by tianon** [comment]
Ah, makes sense; thanks @ChristophBerg!! :heart:

Sticking with 6.0 via this manner makes perfect sense then, and if we have to jump through extra hoops for 7 later, so be it! (good to know that might come :heart:)
